File: leetcode_go/11.Container_With_Most_Water.go
// https://leetcode.com/problems/container-with-most-water/submissions/

func maxArea(height []int) int {
	left, right, res := 0, len(height)-1, 0

	for left < right {
		w := right - left
		h := 0

		if height[left] <= height[right] {
			h = height[left]
			left++
		} else {
			h = height[right]
			right--
		}

		if w*h > res {
			res = w * h
		}
	}

	return res
}

File: leetcode_java/11.Container_With_Most_Water.java
// https://leetcode.com/problems/container-with-most-water/submissions/

class Solution {
  public int maxArea(int[] height) {
    int left = 0;
    int right = height.length - 1;

    int res = 0;

    while (left < right) {
      int w = right - left;
      int h = Math.min(height[left], height[right]);
      res = Math.max(res, w * h);

      if (height[left] <= height[right]) {
        left++;
      } else {
        right--;
      }
    }

    return res;
  }
}